core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,388 @@
+ * @param sendActiveAck a function sending the activation via active ack
+ * @param storeActivation a function storing the activation in a persistent store
+ */
+class ContainerProxy(
+    factory: (TransactionId, String, Exec, ByteSize) => Future[Container],
+    sendActiveAck: (TransactionId, WhiskActivation) => Future[Any],
+    storeActivation: (TransactionId, WhiskActivation) => Future[Any]) extends FSM[ContainerState, ContainerData] with Stash {
+    implicit val ec = context.system.dispatcher
+
+    val unusedTimeout = 30.seconds
+    val pauseGrace = 1.second
+
+    startWith(Uninitialized, NoData())
+
+    when(Uninitialized) {
+        // pre warm a container
+        case Event(job: Start, _) =>
+            factory(
+                TransactionId.invokerWarmup,
+                ContainerProxy.containerName("prewarm", job.exec.kind),
+                job.exec,
+                256.MB)
+                .map(container => PreWarmedData(container, job.exec.kind))
+                .pipeTo(self)
+
+            goto(Starting)
+
+        // cold start
+        case Event(job: Run, _) =>
+            implicit val transid = job.msg.transid
+            factory(
+                job.msg.transid,
+                ContainerProxy.containerName(job.msg.user.namespace.name, job.action.name.name),
+                job.action.exec,
+                job.action.limits.memory.megabytes.MB)
+                .andThen {
+                    case Success(container) => self ! PreWarmedData(container, job.action.exec.kind)
+                    case Failure(t) =>
+                        val response = t match {
+                            case WhiskContainerStartupError(msg) => ActivationResponse.whiskError(msg)
+                            case BlackboxStartupError(msg)       => ActivationResponse.applicationError(msg)
+                            case _                               => ActivationResponse.whiskError(t.getMessage)
+                        }
+                        val activation = ContainerProxy.constructWhiskActivation(job, Interval.zero, response)
+                        sendActiveAck(transid, activation)
+                        storeActivation(transid, activation)
+                }
+                .flatMap {
+                    container =>
+                        run(container, job)
+                            .map(_ => WarmedData(container, job.msg.user.namespace, job.action, Instant.now))
+                }.pipeTo(self)
+
+            goto(Running)
+    }
+
+    when(Starting) {
+        // container was successfully obtained
+        case Event(data: PreWarmedData, _) =>
+            context.parent ! NeedWork(data)
+            goto(Started) using data
+
+        // container creation failed
+        case Event(_: FailureMessage, _) =>
+            context.parent ! ContainerRemoved
+            stop()
+
+        case _ => delay
+    }
+
+    when(Started) {
+        case Event(job: Run, data: PreWarmedData) =>
```

Comment:
should `run` be called `initialize[IfNecessary]AndRun`? or maybe a comment here since it looks like you're running a stem cell without initialization of the code (which happens inside `run`).